sparsnas_decode.cpp

```diff
@@ -313,68 +323,150 @@ int run_for_frequencies(FILE *f, FILE *logfile, float F1, float F2) {
+    if (argc >= 2) {
+        f = fopen(argv[1], "rb");
+        if (!f) {
+            fprintf(stderr, "Failed load!\n");
+            return 1;
         }
-      }
+    }
 
-      if (best_f1 == invalid_f1) {
-        fprintf(stderr, "Nothing found...\n");
-        return 0;
-      }
 
-      range_min = best_f1 - step * 0.5f;
-      range_max = best_f1 + step * 0.5f;
-      step /= 10.0f;
-    } while (step >= 10.0f);
+    //Get SENSOR_ID from environment
+    if (get_env_int("SPARSNAS_SENSOR_ID",&tmp))
+        SENSOR_ID = tmp;
+    else {
+        fprintf(stderr, "SPARSNAS_SENSOR_ID not defined or incorrect. Aborting!\n");
+        return 1;
+    }
 
-    fprintf(stderr, "#define FREQUENCIES {%f, %f}\n", best_f1, best_f1 + 40000.0f);
-  }
+    if (argc >= 3 && strcmp(argv[2], "--find-frequencies") == 0)
+      return run_calibration(f);
+
+
+    //Get the parameters from environment
+    if (get_env_int("SPARSNAS_PULSES_PER_KWH",&tmp))
+        PULSES_PER_KWH = tmp;
+
+    if (get_env_int("SPARSNAS_FREQ_MIN",&tmp))
+        frequencies[0] = tmp;
+    else {
+        fprintf(stderr, "SPARSNAS_FREQ_MIN not defined or incorrect. Aborting!\n");
+        return 1;
+    }
+
+    if (get_env_int("SPARSNAS_FREQ_MAX",&tmp))
+        frequencies[1] = tmp;
+    else {
+        fprintf(stderr, "SPARSNAS_FREQ_MAX not defined or incorrect. Aborting!\n");
+        return 1;
+    }
+
+    FILE *logfile;
+    if (const char *env_p = std::getenv("SPARSNAS_LOG"))
+        logfile=fopen(env_p,"a");
+    else
+        logfile = NULL;
+
+    memset(MQTT_HOSTNAME, '\0', sizeof(MQTT_HOSTNAME));
+    if (const char *env_p = std::getenv("MQTT_HOST"))
+      strncpy(MQTT_HOSTNAME, env_p, sizeof(MQTT_HOSTNAME)-1);
+    else
+      strncpy(MQTT_HOSTNAME, "localhost", sizeof(MQTT_HOSTNAME)-1);
+
+    if (get_env_int("MQTT_PORT",&tmp))
+        MQTT_PORT = tmp;
+
+    memset(MQTT_TOPIC, '\0', sizeof(MQTT_TOPIC));
+    if (const char *env_p = std::getenv("MQTT_TOPIC"))
+      strncpy(MQTT_TOPIC, env_p, sizeof(MQTT_TOPIC)-1);
+    else
+      sprintf(MQTT_TOPIC, "sparsnas/%d", SENSOR_ID);
+
+    memset(MQTT_USERNAME, '\0', sizeof(MQTT_USERNAME));
+    if (const char *env_p = std::getenv("MQTT_USERNAME"))
+      strncpy(MQTT_USERNAME, env_p, sizeof(MQTT_USERNAME)-1);
+
+    memset(MQTT_PASSWORD, '\0', sizeof(MQTT_PASSWORD));
+    if (const char *env_p = std::getenv("MQTT_PASSWORD"))
+      strncpy(MQTT_PASSWORD, env_p, sizeof(MQTT_PASSWORD)-1);
+
+    // Initialize the Mosquitto library
+    mosquitto_lib_init();
+
+    // Create a new Mosquitto runtime instance with a random client ID,
+    mosq = mosquitto_new (NULL, true, NULL);
+
+    if (mosq) {
+      //Set username and password (will be ignored of MQTT_USERNAME=NULL)
+      mosquitto_username_pw_set (mosq, MQTT_USERNAME, MQTT_PASSWORD);
+      int ret = mosquitto_connect (mosq, MQTT_HOSTNAME, MQTT_PORT, 30);
+      if (ret) {
+        fprintf (stderr, "Can't connect to Mosquitto server\n");
+        mosq = NULL;
```

Comment:
@tubalainen What should be the strategy here, should we simply ignore mqtt if a connection can't be established or should we try for infinity (at line  [204](https://github.com/tubalainen/sparsnas_decoder/pull/7/files#diff-b28cdbbd89c121bd0d0cc5d9e03116d1R204))?
Another option could be to let the decoder die here if a mqtt connection can't be established..